### ISSUE TITLE
Stackoverflows and anomalies with certain notations

### ISSUE THREAD

**Issue by LasseBlaauwbroek** [coq/coq#13078]
#### Description of the problem

I'm not sure if these are related, but here are three slight modifications of a snippet. At least the last two give unexpected results. The last one, I am encountering anywhere Coq is trying to print a notation in a pattern match. Also, `Fail` seems to be unable to catch anomalies?

```
Notation "x &" := (Some x)
  (at level 0, x ident).

Example t (x : option unit) := match x with end.
(* Unexpected but probably not a bug *)
(* Syntax error: '&' expected after [Prim.name] (in [constr:operconstr]).*)
```
```
Notation "x &" := (Some x)
  (at level 0, x pattern).

Example t (x : option unit) := match x with end.
(* Stackoverflow *)
```
```
Declare Custom Entry test.

Notation "x &" := (Some x)
  (in custom test at level 0, x pattern).

Definition t (x : option unit) := match x with | None => None | Some tt => Some tt end.

Fail Example u (x : option unit) := match x with end.
(* <in exception printer>:<original exception:Anomaly
                                           "File "interp/notation_ops.ml", line 1437, characters 27-33: Assertion failed."
Please report at http://coq.inria.fr/bugs/. *)

Goal forall x, t x = None.
Fail Fail unfold t.
(* The command has indeed failed with message:
The command has not failed! *)
unfold t.
(* Anomaly "File "interp/notation_ops.ml", line 1437, characters 27-33: Assertion failed."
Please report at http://coq.inria.fr/bugs/. *)
```
#### Coq Version
`master`

**Comment by herbelin** [comment]
Hi and thanks for the report. Here is an analysis of the failures:

> ```
> Notation "x &" := (Some x)
>   (at level 0, x ident).
> 
> Example t (x : option unit) := match x with end.
> (* Unexpected but probably not a bug *)
> (* Syntax error: '&' expected after [Prim.name] (in [constr:operconstr]).*)
> ```

The added rule is not factorized with the rule parsing `ident` (as shown by `Print Grammar constr`). First, `ident` in `constr` is actually bound to `name` (see ongoing #11841). Second, identifiers are parsed with an indirection in the `atomic_constr` subentry. This kind of `constr` extension you'd like currently goes beyond the scope of what notations allow (though it is worth to think at whether better can be done).

> ```
> Notation "x &" := (Some x)
>   (at level 0, x pattern).
> 
> Example t (x : option unit) := match x with end.
> (* Stackoverflow *)
> ```

This introduces an indirection from `constr` to the `pattern` grammar and a loop inside the `pattern` grammar (as shown by `Print Grammar pattern`). This explains the stack overflow. Indeed, for every addition to the `constr`  grammar, an addition is made to the `pattern` grammar. The meaning of such addition when the `constr` notation refers to binders or `pattern` is questionable. I will think about it.

> ```
> Declare Custom Entry test.
> 
> Notation "x &" := (Some x)
>   (in custom test at level 0, x pattern).
> 
> Definition t (x : option unit) := match x with | None => None | Some tt => Some tt end.
> 
> Fail Example u (x : option unit) := match x with end.
> (* <in exception printer>:<original exception:Anomaly
>                                            "File "interp/notation_ops.ml", line 1437, characters 27-33: Assertion failed."
> Please report at http://coq.inria.fr/bugs/. *)

The anomaly is raised by the printer. The cause is similar to the previous loop. The rule is added to  the `pattern` part of custom entry `test` but the printer was not expecting that a `pattern` would occur here.

For the record, the same anomaly can be triggered without custom notations:
```
Notation "x &" := (Some x)
  (at level 0, x pattern).
Check fun x => match x with Some _ => true | _ => false end. (* anomaly *)
```

> ```
> Goal forall x, t x = None.
> Fail Fail unfold t.
> (* The command has indeed failed with message:
> The command has not failed! *)
> unfold t.
> (* Anomaly "File "interp/notation_ops.ml", line 1437, characters 27-33: Assertion failed."
> Please report at http://coq.inria.fr/bugs/. *)
> ```
This is intended. An anomaly is considered to be a Coq misbehavior and not a user failure.

**Comment by herbelin** [comment]
By the way, the idea to support the vector notation discussed in the [LL(n)](https://coq.discourse.group/t/ll-n-notations-in-coq/1057) discourse topic is nice and I did some experiments (though without full success yet).

First, supporting empty productions in custom entries seems indeed to make sense. I'll think at it.

Second, once empty productions are made available, I'm not desperating to arrive to a solution by using a different presentation of the `S` language. We would then exploit the look-ahead over tokens made by the underlying Camlp5 parsing engine. That is, in the presence of competing rules such as `ident & T` and `ident U`, the presence of `&` would be detected before entering the first rule with a fallback on the second rule if `&` not present. Here is an example of application of this look-ahead:
```
Declare Custom Entry S.
Notation "[ x ]" := x (x custom S).

Notation "x & c" := (0,x,c) (in custom S, x ident, c constr at level 0).
Notation "x c" := (1,x,c) (in custom S, x ident, c constr at level 0).

Check fun x => [x & 1]. (* ok *)
Check fun x => [x 1]. (* ok *)
```

However, I don't feel confident enough on the exact behavior of the internal of Camlp5 to ensure that it will work before having tested. So, the first step would be to add support for epsilon rules.

**Comment by LasseBlaauwbroek** [comment]
Hi @herbelin. Thanks for the analysis. For the record, I'm only encountering the printing issue in the "wild" (the others I found while I was diagnosing this).

Regarding the LL(n) topic: I actually discovered that the grammar I am using can be converted to LL(1) (not the one I presented in the topic though, that is a school-example of LL(2)). It just becomes very verbose, because I had to factorize a lot of grammar rules. So the "vector" idea is currently working for me, modulo this printing ussue. But I still expect that it is easy to end up with a non-LL(1) grammar if you want to use the unicode-trick.

Regarding empty productions: These are actually working for me in some situations but not in others. For example, they do work here:
```
Declare Custom Entry test.
Notation "[ x ]" := x (x custom test).
Notation "" := tt (in custom test).
Check [].
```
But in more complicated grammars it tends to be hit and miss.

One suggestion: Maybe it could be a good idea to explicitly support unicode combinators through a separate command? This may be easier and more robust to implement.
